cdap-app-fabric/src/main/java/io/cdap/cdap/app/deploy/AppDeletionSubscriberService.java

```diff
@@ -0,0 +1,144 @@
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.TableNotFoundException;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import org.apache.tephra.TxConstants;
+
+import java.io.IOException;
+import java.util.Iterator;
+import javax.annotation.Nullable;
+
+/**
+ * A TMS Subscriber Service responsible for consuming app deletion messages from TMS and executes removal from
+ * the app spec table. Subscribes to the Constants.AppFabric.APP_DELETION_EVENT_TOPIC topic.
+ */
+public class AppDeletionSubscriberService extends AbstractMessagingSubscriberService<AppDeletionMessage> {
+  private static final String SUBSCRIBER = "appdelete.writer";
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
+    .create();
+
+  private final CConfiguration cConf;
+  private final MultiThreadMessagingContext messagingContext;
+  private final TransactionRunner transactionRunner;
+  private final int maxRetriesOnConflict;
+  private final MetricsCollectionService metricsCollectionService;
+
+
+  @Inject
+  AppDeletionSubscriberService(CConfiguration cConf, MessagingService messagingService,
+                            MetricsCollectionService metricsCollectionService,
+                            TransactionRunner transactionRunner) {
+    super(
+      NamespaceId.SYSTEM.topic(cConf.get(Constants.AppFabric.APP_DELETION_EVENT_TOPIC)),
+      cConf.getInt(Constants.AppFabric.APP_DELETE_EVENT_FETCH_SIZE),
+      cConf.getInt(TxConstants.Manager.CFG_TX_TIMEOUT),
+      cConf.getLong(Constants.AppFabric.APP_DELETE_EVENT_POLL_DELAY_MILLIS),
+      RetryStrategies.fromConfiguration(cConf, "system.appdelete."),
+      metricsCollectionService.getContext(ImmutableMap.of(
+        Constants.Metrics.Tag.COMPONENT, Constants.Service.MASTER_SERVICES,
+        Constants.Metrics.Tag.INSTANCE_ID, "0",
+        Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+        Constants.Metrics.Tag.TOPIC, cConf.get(Constants.AppFabric.APP_DELETION_EVENT_TOPIC),
+        Constants.Metrics.Tag.CONSUMER, SUBSCRIBER
+      )));
+
+    this.cConf = cConf;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactionRunner = transactionRunner;
+    this.maxRetriesOnConflict = cConf.getInt(Constants.Metadata.MESSAGING_RETRIES_ON_CONFLICT);
+    this.metricsCollectionService = metricsCollectionService;
+  }
+
+  @Override
+  protected TransactionRunner getTransactionRunner() {
+    return transactionRunner;
+  }
+
+  @Nullable
+  @Override
+  protected String loadMessageId(StructuredTableContext context) throws Exception {
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+    return appMetadataStore.retrieveSubscriberState(getTopicId().getTopic(), SUBSCRIBER);
+  }
+
+  @Override
+  protected void storeMessageId(StructuredTableContext context, String messageId) throws Exception {
```

Comment:
That's tons of boilerplate code copied between a lot of classes. Could you please extract an abstract class?

Reply:
Okay let me create a separate abstract class